google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudLibraryProjectState.java

```diff
@@ -0,0 +1,164 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.apis;
+
+import com.google.cloud.tools.libraries.json.CloudLibrary;
+import com.google.common.collect.ImmutableSet;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.components.ProjectComponent;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Computable;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.fest.util.Maps;
+import org.fest.util.VisibleForTesting;
+import org.jetbrains.idea.maven.dom.MavenDomUtil;
+import org.jetbrains.idea.maven.dom.model.MavenDomDependency;
+import org.jetbrains.idea.maven.dom.model.MavenDomProjectModel;
+import org.jetbrains.idea.maven.project.MavenProject;
+import org.jetbrains.idea.maven.project.MavenProjectsManager;
+import org.jetbrains.idea.maven.project.MavenProjectsManager.Listener;
+
+/**
+ * A {@link ProjectComponent} that maintains, in memory, the current set of {@link CloudLibrary}
+ * that are configured on the user's project.
+ */
+public class CloudLibraryProjectState implements ProjectComponent {
+
+  private static final Logger logger = Logger.getInstance(AddCloudLibrariesAction.class);
+
+  private final Project project;
+  private Map<Module, Set<CloudLibrary>> moduleLibraryMap = Maps.newHashMap();
+  private List<CloudLibrary> allLibraries;
+
+  private CloudLibraryProjectState(Project project) {
+    this.project = project;
+  }
+
+  static CloudLibraryProjectState getInstance(Project project) {
+    return project.getComponent(CloudLibraryProjectState.class);
+  }
+
+  /**
+   * Attaches a listener to listen for changes to a module's pom.xml to request a resync of the
+   * project's managed dependencies.
+   *
+   * <p>If the project is not a maven project, then the listener simply isn't fired. If the project
+   * is not a maven project, but later becomes one, then the listener will be fired as expected and
+   * the managed libraries will be synchronized.
+   */
+  @Override
+  public void projectOpened() {
+    // todo (eshaul) handle Gradle & native projects and ensure that double syncing won't occur
+    MavenProjectsManager.getInstance(project)
+        .addManagerListener(
+            new Listener() {
+              @Override
+              public void projectsScheduled() {
+                syncManagedProjectLibraries();
+              }
+            });
+  }
+
+  /**
+   * Returns an optional set of {@link CloudLibrary} currently configured on the given {@link
+   * Module}.
+   */
+  Set<CloudLibrary> getManagedLibraries(Module module) {
```

Comment:
Looks like for now this method is only used in the tests, what is its further target, do we plan to maintain state of cloud libraries for a project depending on their version compatibility? I.e. for now it seems we derive this mapping from POM, the next step would be to enforce specific versions?

Reply:
correct. this is the main entry point into the state of the managed libraries for a given module. It will be used to control versions, like you said. Also for BOM stuff in the UI (where we are scoped to a specific module).

Reply:
and yeah its only in tests now, but will be used in all those areas in subsequent PRs

Reply:
So when we add library version control, we won't be able to cleanly derive this mapping from POM itself, and that's the reason we will keep this as a separate mapping?

Reply:
sorry not sure I follow.

when a library is added (either via our UI and indirectly to the pom) or directly to the pom, then the maven project listener will be fired and force a resynchronization of this mapping. This sync will be called at the same time the user does a manual maven sync (or automatic if they have that setting enabled). This way, our knowledge of the state of the managed libraries is also in sync with the maven integration itself.

I'm not sure if I addressed your question.

Reply:
Oh got it, we just update it on every listener update, I confused that with some cached state different from pom.xml file state. It's clear, thanks!